clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java

```diff
@@ -436,4 +442,71 @@ static final String firstPrincipal(Subject subject) {
                 throw new KafkaException("Principal could not be determined from Subject, this may be a transient failure due to Kerberos re-login");
         }
     }
+
+    /**
+     * Kerberos exceptions that may require special handling. The standard Kerberos error codes
+     * for these errors are retrieved using KrbException#errorCode() from the underlying Kerberos
+     * exception thrown during {@link SaslClient#evaluateChallenge(byte[])}.
+     */
+    private enum KerberosError {
+        // (Mechanism level: Server not found in Kerberos database (7) - UNKNOWN_SERVER)
+        SERVER_NOT_FOUND(7, false),
```

Comment:
Is it necessary to handle non-retriable error like this explicitly?

Reply:
No, I included this one because we were adding additional logging for this case, but were identifying this error using the exception string. I modified that to use the same error detection. Added a comment.